text/0000-at-least-one-feature.md

```diff
@@ -0,0 +1,183 @@
+Allow packages to require that dependencies on them must specify at least one feature (the `default` feature counts).
+This avoids backwards compatibility problems with `default-features = false`.
+
+# Motivation
+[motivation]: #motivation
+
+A major use-case of Cargo features to take previously mandatory functionality and make it optional.
+This is usual done in order to make the code more portable than it was previously, while not breaking existing consumers of the library.
+Consider this example which shoes both work works and what doesn't:
+
+1. Library has no features.
+
+2. A `foo` feature is added, gating functionality that already existed.
+   It is on by default.
+   `no-default-features = true` can be used in which some dependencies only needed for `foo` can be avoided.
+   Yay!
+
+3. A `bar` feature is added, gating functionality that already existed.
+
+   - Suppose it is off by default.
+     Oh no!
+     All Existing use-cases break because the functionality that depends on `bar` goes away.
+
+   - Suppose it is on by default, or depended-upon by `bar`.
+     Oh no!
+     Existing `no-default-features = true` are now broken!
+     They want a feature set of `{bar}` which would correspond to the old `{}`, but there is no way to arrange it.
+
+In step two, we could "ret-con" the empty feature set with the `default` feature.
+But this is a trick that can only be pulled once.
+The second time around, we already have a default feature; we are out of luck.
+
+The previous attempts attempted to make new default features, or migrate the requested feature sets.
+But that is complex.
+There is exactly a simpler solution: simply require that *some* feature always be depended-upon.
+
+To see why this works, it helps to first see that step 2 was *already* broken.
+Here's the thing, even though there previously were not any features, that *doesn't* mean there were not any `no-default-features = true` users!
+Sure, it wouldn't do anything for crate with new features, but one can still use it.
+Then when just `bar` is added, we already have a problem, because the `default` feature will no "catch" all the existing users ---
+the mischievous users that were already using `no-default-features = true` will have their code broken!
+
+This brings us to the heart of the problem.
+So long as users are depending on "something", we can be careful to make sure those features keep their meaning.
+But when users are depending on nothing at all with `no-default-features = true` and an empty feature set, we have nothing to "hook into".
+The simple solution is just to rule out that problem entirely!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Packages with
+```toml
+[package]
+at-least-one-feature = true
+```
+are easier to maintain!
+You don't need to worry about the `no-default-features = true, features = []` case anymore.
+You can be sure that all consumers must dependent on either `default` or a regular named feature.
```

Comment:
Isn't adding this a breaking change? My concern is people won't plan for needing to add this but the point of these RFCs is to avoid breaking changes.

Reply:
See this thread above https://github.com/rust-lang/rfcs/pull/3347#discussion_r1020616763. It will prevent people from upgrading but I do not consider it a breaking change because there is no risk of "incorrect plans". I will add this to the RFC.

Reply:
Closing this out in favor of that conversation.